### PR TITLE
Fix issues with ARRAY columns and NULL values

### PR DESCRIPTION
**Pull request by RobertBrunel**
This change fixes various issues related to the nullability of ARRAYs that can surface in INSERT/UPDATE statements against such columns.

* In `MessageHelper#coerceObject()`, add code to correctly wrap the evaluated array into the appropriate wrapper message when the target is a nullable array. Before, it would return the array value as-is as a `List` object, which later causes a protobuf type mismatch in `FieldSet$Builder#verifyType()`. This issue surfaces in SQL when trying to INSERT a NULL array into a nullable ARRAY column.

* In `MessageHelper#hasAnySuchField()`, change the code so that a repeated field is reported as present even when it has 0 repetitions. (See the explaining comment there.) This fixes INSERTing an empty array via `VALUES ([])`, where the [] would incorrectly get mapped to NULL by `MessageHelper#coerceMessage()`.

* In `NullableArrayTypeUtils#unwrapIfArray()`, skip unwrapping the wrapper message in the case where the `wrappedValue` is not actually a `Message` but already a `List`. Without this, a simple SQL UPDATE query against an ARRAY NOT NULL column runs into `ClassCastException`.

* In `RecordConstructorValue#eval()`, improve the message of the verify statement that prevents assigning a NULL array to a non-nullable field as in `INSERT INTO … (…, not_null_array) VALUES (…, NULL)`.

Testing:
* Add small regression tests to `arrays.yamsql` for the three items.

This fixes #4069.

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/NullableArrayTypeUtils.java [hunk 1]
```diff
@@ -105,7 +105,7 @@ public static Object unwrapIfArray(@Nullable Object wrappedValue, @Nonnull Type
         // If the last step in the field path is an array that is also nullable, then we need to unwrap the value
         // wrapper.
         //
-        if (wrappedValue != null && type.isArray() && type.isNullable()) {
+        if (wrappedValue != null && wrappedValue instanceof Message && type.isArray() && type.isNullable()) {
```
Comment (pengpeng-lu) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/NullableArrayTypeUtils.java]: In what scenario do we expect wrappedValue not be instanceof Message? And in this scenario what should we return?

File: yaml-tests/src/test/resources/arrays.yamsql [hunk 2]
```diff
@@ -19,58 +19,93 @@
...
+      # INSERT for an ARRAY of a multi-element STRUCT type.
       - query: INSERT INTO E VALUES (1, [(1, 1.0), (2, 2.0), (3, 3.0)]), (2, [(2, 2.0), (3, 3.0), (4, 4.0)]), (3, [(3, 3.0), (4, 4.0), (5, 5.0)])
       - count: 3
     -
+      # INSERT for an ARRAY of a multi-element STRUCT type featuring an INTEGER ARRAY member.
       - query: INSERT INTO F VALUES
               (1, [(1, 1.0, [1, 1, 1]), (2, 2.0, [1, 1, 1]), (3, 3.0, [1, 1, 1])]),
               (2, [(2, 2.0, [1, 1, 1]), (3, 3.0, [1, 1, 1]), (4, 4.0, [1, 1, 1])]),
               (3, [(3, 3.0, [1, 1, 1]), (4, 4.0, [1, 1, 1]), (5, 5.0, [1, 1, 1])])
       - count: 3
```
Comment (pengpeng-lu) [inline yaml-tests/src/test/resources/arrays.yamsql:124]: is array x and array x.z nullable or non-nullable? Could you add some tests to insert `null` values into arrays in all the tables (not just table A)?

**Review by pengpeng-lu — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/NullableArrayTypeUtils.java (hunk 1 above)
Comment (RobertBrunel) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/NullableArrayTypeUtils.java]: `unwrapIfArray()` gets called from only two places.

One is `RecordQueryAbstractDataModificationPlan#mutateRecord()` › `transformMessage()`. In this code path the `wrappedValue` is a `Message`.

The other is when `unwrapIfArray()` gets called from `FieldValue#eval()`. In that code path the array is already an `ArrayList`, not a `Message`. In this scenario just returning the `ArrayList` as-is is correct.

The new test case that triggers this case is:

```
# UPDATE for a not-nullable INTEGER ARRAY.
- query: UPDATE A_NOT_NULL SET X = [11, 12, 13] WHERE PK = 1 RETURNING "new".X
```

It seems that until now we haven’t had any test that tries to `UPDATE` a `NOT NULL` array to a new value. So this scenario went unnoticed.

I’m clarifying this a bit in the comments.

File: yaml-tests/src/test/resources/arrays.yamsql (hunk 2 above)
Comment (RobertBrunel) [reply yaml-tests/src/test/resources/arrays.yamsql:124]: I’ve added more tests, also for the UPDATE-to-NULL and UPDATE-to-[] cases.

**Review by pengpeng-lu — APPROVED** [review]
LGTM!
